File: Front.java
import javafx.application.Application;
import javafx.application.Platform;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;
import javafx.concurrent.Task;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

public class Front extends Application {

    private Label qoeBeforeLabel;
    private Label qoeAfterLabel;

    @Override
    public void start(Stage primaryStage) {
        Button startButton = new Button("开始抓包");
        qoeBeforeLabel = new Label("流量整形前QoE结果: ");
        qoeAfterLabel = new Label("流量整形后QoE结果: ");

        startButton.setOnAction(event -> startCapture());

        VBox vbox = new VBox(10, startButton, qoeBeforeLabel, qoeAfterLabel);
        Scene scene = new Scene(vbox, 300, 200);
        primaryStage.setScene(scene);
        primaryStage.setTitle("Dash评估系统");
        primaryStage.show();
    }

    private void startCapture() {
        Task<Void> captureTask = new Task<Void>() {
            @Override
            protected Void call() throws Exception {
                // 使用 tcpdump 抓包并保存为 output.pcap
                String command = "tcpdump -i any -w output.pcap"; // 根据实际情况修改接口
                Process process = Runtime.getRuntime().exec(command);
                // 等待一段时间（根据实际情况调整）
                Thread.sleep(5000); // 假设抓包持续5秒
                process.destroy(); // 停止 tcpdump
                return null;
            }

            @Override
            protected void succeeded() {
                // 抓包完成后获取 QoE 结果
                fetchQoEResults();
            }
        };
        new Thread(captureTask).start();
    }

    private void fetchQoEResults() {
        Task<String[]> qoeTask = new Task<String[]>() {
            @Override
            protected String[] call() throws Exception {
                String[] results = new String[2];
                // 假设你的后端 Python 服务在 localhost:5000 上
                results[0] = sendRequest("http://localhost:5000/getQoEResults?type=before");
                results[1] = sendRequest("http://localhost:5000/getQoEResults?type=after");
                return results;
            }
        };

        qoeTask.setOnSucceeded(event -> {
            String[] results = qoeTask.getValue();
            Platform.runLater(() -> {
                qoeBeforeLabel.setText("流量整形前QoE结果: " + results[0]);
                qoeAfterLabel.setText("流量整形后QoE结果: " + results[1]);
            });
        });

        new Thread(qoeTask).start();
    }

    private String sendRequest(String urlString) {
        try {
            URL url = new URL(urlString);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");
            if (conn.getResponseCode() == 200) {
                StringBuilder response = new StringBuilder();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        response.append(line);
                    }
                }
                return response.toString();
            } else {
                return "请求失败，状态码: " + conn.getResponseCode();
            }
        } catch (IOException e) {
            return "请求异常: " + e.getMessage();
        }
    }

    public static void main(String[] args) {
        launch(args);
    }
}

File: Front.py
import tkinter as tk
from tkinter import ttk
import subprocess
import requests
import threading
import time
import json

class FrontApp:
    def __init__(self, root):
        self.root = root
        self.root.title("Dash评估系统")

        self.qoe_before_label = ttk.Label(root, text="流量整形前QoE结果: ")
        self.qoe_after_label = ttk.Label(root, text="流量整形后QoE结果: ")

        self.start_button = ttk.Button(root, text="开始抓包", command=self.start_capture)

        self.start_button.pack(pady=10)
        self.qoe_before_label.pack(pady=10)
        self.qoe_after_label.pack(pady=10)

    def start_capture(self):
        capture_thread = threading.Thread(target=self.capture_task)
        capture_thread.start()

    def capture_task(self):
        command = ["tcpdump", "-i", "any", "-w", "output.pcap"]
        process = subprocess.Popen(command)
        time.sleep(5)
        process.terminate()
        process.wait()
        self.fetch_qoe_results()

    def fetch_qoe_results(self):
        qoe_thread = threading.Thread(target=self.qoe_task)
        qoe_thread.start()

    def qoe_task(self):
        try:
            # Adjust the paths and server IP as necessary
            results = [self.send_request("http://localhost:5000/qoe", "/home/luo/Desktop/test01.pcap", "192.168.5.130", "before"),
                       self.send_request("http://localhost:5000/qoe", "/home/luo/Desktop/test01.pcap", "192.168.5.130", "after")]
            self.root.after(0, self.update_labels, results)
        except Exception as e:
            print(f"请求异常: {e}")

    def send_request(self, url, pcap_file, server_ip, qoe_type):
        headers = {'Content-Type': 'application/json'}
        data = {'pcap_file': pcap_file, 'server_ip': server_ip, 'type': qoe_type}
        response = requests.post(url, headers=headers, data=json.dumps(data))
        if response.status_code == 200:
            return response.json().get("qoe_score", "N/A")
        else:
            return f"请求失败，状态码: {response.status_code}"

    def update_labels(self, results):
        self.qoe_before_label.config(text=f"流量整形前QoE结果: {results[0]}")
        self.qoe_after_label.config(text=f"流量整形后QoE结果: {results[1]}")

if __name__ == "__main__":
    root = tk.Tk()
    app = FrontApp(root)
    root.mainloop()